File: Data Structures & Algorithms/top-k-elements-in-list/submission-2.py
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        counter = defaultdict(int)
        for n in nums:
            counter[n] = counter[n] + 1
        top = sorted(counter.items(), key=lambda item: item[1])
        result = []
        for idx in range(1, k + 1):
            result.append(top[-idx][0])
        return result
        
        # top_elements = [0] * k
        # top_freq   = [-1001] * k
        # counter.sort(counter.items, )
        #     for idx, freq in enumerate(top_freq):
        #         if counter[n] > freq:
        #             top_elements[idx], top_freq[idx] = n, counter[n]
        #             break
        # return top_elements
        

File: Data Structures & Algorithms/products-of-array-discluding-self/submission-2.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        final int size = nums.length;
        
        final int[] prefix = new int[size];
        int prev = 1;
        for (int i = 0; i < size; i++) {
            prefix[i] = prev;
            prev *= nums[i];
        }

        final int[] suffix = new int[size];
        prev = 1;
        for (int i = size - 1; i >= 0; i--) {
            suffix[i] = prev;
            prev *= nums[i];
        }

        final int[] result = new int[size];
        for (int i = 0; i < size; i++) {
            result[i] = prefix[i] * suffix[i];
        }
        return result;
    }
}  
